File: backend/src/snowan/server/system.py
"""Tools catalog, app preferences, and about/diagnostics."""
import subprocess

from fastapi import APIRouter
from pydantic import BaseModel

from .. import audit, config
from ..agent.build import tool_catalog

router = APIRouter()

APP_VERSION = "0.1.0"


class PrefsPatch(BaseModel):
    approval_mode: str | None = None
    max_iters: int | None = None
    timezone: str | None = None
    auto_title: bool | None = None
    system_prompt: str | None = None
    profile: dict | None = None
    web_search_provider: str | None = None
    tavily_api_key: str | None = None
    brave_api_key: str | None = None
    jina_api_key: str | None = None
    disabled_tools: list[str] | None = None


@router.get("/api/tools")
def list_tools() -> list[dict]:
    return tool_catalog()


@router.get("/api/audit")
def audit_log(limit: int = 100) -> list[dict]:
    """Recent tool calls (newest first) for the 权限 audit view."""
    return audit.recent(limit)


@router.get("/api/prefs")
def get_prefs() -> dict:
    return config.load_prefs()


@router.put("/api/prefs")
def put_prefs(patch: PrefsPatch) -> dict:
    return config.save_prefs(patch.model_dump(exclude_none=True))


@router.get("/api/about")
def about() -> dict:
    s = config.load_settings()
    return {
        "version": APP_VERSION,
        "data_dir": str(config.SNOWAN_HOME),
        "provider": s.provider,
        "model": s.model,
        "has_api_key": bool(s.api_key),
    }


@router.post("/api/about/open-data-dir")
def open_data_dir() -> dict:
    config.SNOWAN_HOME.mkdir(parents=True, exist_ok=True)
    subprocess.Popen(["open", str(config.SNOWAN_HOME)])
    return {"ok": True}

File: frontend/src/api/system.ts
import { api } from './base';

export type ToolInfo = { name: string; description: string; mutating: boolean; enabled: boolean };
export type AuditEntry = { ts: string; tool: string; summary: string; status: string };
export type Profile = { name: string; location: string; notes: string };
export type Prefs = {
  approval_mode: string; // 'auto' | 'ask' | 'strict'
  max_iters: number;
  timezone: string;
  auto_title: boolean;
  system_prompt: string;
  profile: Profile;
  web_search_provider: string; // 'duckduckgo' | 'tavily' | 'brave'
  tavily_api_key: string;
  brave_api_key: string;
  jina_api_key: string;
  disabled_tools: string[];
  memory_enabled: boolean;
  onboarded: boolean;
};
export type About = {
  version: string;
  data_dir: string;
  provider: string;
  model: string;
  has_api_key: boolean;
};
export type Usage = {
  turns: number;
  input: number;
  output: number;
  cache_read: number;
  cache_write: number;
  hit_rate: number;
};

const j = <T>(r: Response): Promise<T> => {
  if (!r.ok) throw new Error(`${r.status}`);
  return r.json();
};

export const getTools = () => fetch(api('/api/tools')).then((r) => j<ToolInfo[]>(r));
export const getAudit = (limit = 100) =>
  fetch(api(`/api/audit?limit=${limit}`)).then((r) => j<AuditEntry[]>(r));
export const getPrefs = () => fetch(api('/api/prefs')).then((r) => j<Prefs>(r));
export const savePrefs = (patch: Partial<Prefs>) =>
  fetch(api('/api/prefs'), {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(patch),
  }).then((r) => j<Prefs>(r));
export const getAbout = () => fetch(api('/api/about')).then((r) => j<About>(r));
export const getUsage = () => fetch(api('/api/usage')).then((r) => j<Usage>(r));
export const openDataDir = () => fetch(api('/api/about/open-data-dir'), { method: 'POST' });
